### PR TITLE
Limit decoding attribute to images with src starting with double quote

### PR DESCRIPTION
**Pull request by adamsilverstein**
Trac ticket: https://core.trac.wordpress.org/ticket/56969

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/media.php [hunk 1]
```diff
@@ -3162,6 +3162,31 @@ public function test_wp_img_tag_add_loading_attr_opt_out() {
...
 
+
+
+	/**
+	 * Test that decoding="async" is not applied to img tags with single quotes.
+	 *
+	 * @ticket 56969
+	 */
+	public function test_wp_img_tag_add_decoding_attr_with_single_quotes() {
+		$img = "<img src='example.png' alt=' width='300' height='225' />";
```
Comment (felixarntz) [inline tests/phpunit/tests/media.php]: Nit-pick, but the HTML here is broken. Doesn't matter for the test but probably good to use a working "realistic" example.
```suggestion
		$img = "<img src='example.png' alt='' width='300' height='225' />";
```

File: tests/phpunit/tests/media.php [hunk 2]
```diff
@@ -3162,6 +3162,31 @@ public function test_wp_img_tag_add_loading_attr_opt_out() {
...
+		$this->assertStringNotContainsString( ' decoding="async"', $img );
+	}
+
+	/**
+	 * Test that decoding="async" os not applied to img tags inside JSON.
+	 *
+	 * @ticket 56969
+	 */
+	public function test_decoding_async_not_applied_to_json() {
+		$content = '{"image": "<img src=\"example.png\" alt=\" width=\"300\" height=\"225\" />"}';
```
Comment (felixarntz) [inline tests/phpunit/tests/media.php]: Same as above.
```suggestion
		$content = '{"image": "<img src=\"example.png\" alt=\"\" width=\"300\" height=\"225\" />"}';
```

**Review by felixarntz — APPROVED** [review]
@adamsilverstein Looks solid to me, just two nit-picks here in the test HTML.

File: tests/phpunit/tests/media.php (hunk 2 above)
Comment (peterwilsoncc) [reply tests/phpunit/tests/media.php]: Another nit: `\` should be escaped as `\\`, even in single quoted strings.

PHP parses it fine as is

**Review by peterwilsoncc — COMMENTED** [review]
This seems Ok as temporary fix but as single quotes are a perfectly legitimate form of HTML, it would be good to follow up with something more elegant.

WordPress itself uses the quote types interchangeably and it seems to be fairly common in plugins too, especially when generating the markup from a set of variables.

https://github.com/WordPress/wordpress-develop/blob/55d26534df75bbd13cc36a9e4e0881c5f8cc223c/src/wp-includes/pluggable.php#L2890

File: src/wp-includes/media.php [hunk 3]
```diff
@@ -1962,6 +1962,13 @@ function wp_img_tag_add_loading_attr( $image, $context ) {
  * @return string Converted `img` tag with `decoding` attribute added.
  */
 function wp_img_tag_add_decoding_attr( $image, $context ) {
+
```
Comment (mukeshpanchal27) [inline src/wp-includes/media.php]:
```suggestion
```

Don't need top space as other `wp_img_tag_add_loading_attr()` also not used with similar starting docblock.

File: tests/phpunit/tests/media.php [hunk 4]
```diff
@@ -3162,6 +3162,31 @@ public function test_wp_img_tag_add_loading_attr_opt_out() {
 		$this->assertStringNotContainsString( ' loading=', $img );
 	}
 
+
+
+	/**
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/media.php]:
```suggestion
	/**
```

File: tests/phpunit/tests/media.php [hunk 5]
```diff
@@ -3162,6 +3162,31 @@ public function test_wp_img_tag_add_loading_attr_opt_out() {
...
+	 */
+	public function test_wp_img_tag_add_decoding_attr_with_single_quotes() {
+		$img = "<img src='example.png' alt=' width='300' height='225' />";
+		$img = wp_img_tag_add_decoding_attr( $img, 'test' );
+
+		$this->assertStringNotContainsString( ' decoding="async"', $img );
+	}
+
+	/**
+	 * Test that decoding="async" os not applied to img tags inside JSON.
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/media.php]:
```suggestion
	 * Test that decoding="async" is not applied to img tags inside JSON.
```

File: tests/phpunit/tests/media.php [hunk 6]
```diff
@@ -3162,6 +3162,31 @@ public function test_wp_img_tag_add_loading_attr_opt_out() {
...
+
+	/**
+	 * Test that decoding="async" is not applied to img tags with single quotes.
+	 *
+	 * @ticket 56969
+	 */
+	public function test_wp_img_tag_add_decoding_attr_with_single_quotes() {
+		$img = "<img src='example.png' alt=' width='300' height='225' />";
+		$img = wp_img_tag_add_decoding_attr( $img, 'test' );
+
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/media.php]:
```suggestion
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @adamsilverstein. It looks good to me left some feedback for space and text changes.

https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/media/wpImageTagAddDecodingAttr.php contains all the unit tests for `wp_img_tag_add_decoding_attr()` so is it better to move this new unit test there?

**Comment by desrosj** [comment]
Applied some of the more straightforward suggestions. @adamsilverstein would you be able to get this committed at some point today?

**Comment by adamsilverstein** [comment]
Yes, will do.

**Comment by adamsilverstein** [comment]
> This seems Ok as temporary fix but as single quotes are a perfectly legitimate form of HTML, it would be good to follow up with something more elegant.
> 
> WordPress itself uses the quote types interchangeably and it seems to be fairly common in plugins too, especially when generating the markup from a set of variables.
> 
> https://github.com/WordPress/wordpress-develop/blob/55d26534df75bbd13cc36a9e4e0881c5f8cc223c/src/wp-includes/pluggable.php#L2890

I agree it would be nice to have a more robust solution; this one seems ok for now - especially since the lazy attribute callback already has an identical check.

**Comment by adamsilverstein** [comment]
>  Applied some of the more straightforward suggestions. @adamsilverstein would you be able to get this committed at some point today?

@desrosj  I'll get this committed and backported first thing tomorrow if you don't beat me to it.

**Comment by peterwilsoncc** [comment]
Merged in https://core.trac.wordpress.org/changeset/54802 / https://github.com/WordPress/wordpress-develop/commit/51032426b54113cf4acccd27de693ab7158a9f92

**Comment by adamsilverstein** [comment]
thanks @peterwilsoncc!

**Pull request closed without merging**